tpp-app/src/services/cycle/CycleService.js

```diff
@@ -0,0 +1,381 @@
+    try{
+      let today = new Date();
+      let today_str = getDateString(today);
+      let percent = await AsyncStorage.getItem(Keys.CYCLE_DONUT_PERCENT)
+      percent = percent != null ? JSON.parse(percent) : null;
+
+
+      let calendar = await getCalendarByYear(today.getFullYear());
+
+      if (percent != null && today_str in percent){
+        return percent[today_str];
+      }
+      else{
+        let mostRecentPeriodStart = await this.GETMostRecentPeriodStartDate(calendar);
+        let avgCycleLength = await this.GETAverageCycleLength(calendar);
+        if (mostRecentPeriodStart && avgCycleLength){
+          let daysSincePeriodStart = getDaysDiffInclusive(mostRecentPeriodStart, today);
+          let cycleDonutPercent = daysSincePeriodStart / avgCycleLength;
+          this.POSTCycleDonutPercent(cycleDonutPercent);
+          return cycleDonutPercent;
+        }
+        else{
+          return 0;
+        }
+      }
+
+    } catch(e){
+      console.log(e);
+
+    }
+  },
+
+  /** Get the start and length of each period in the given year
+   * @param {number} year The year to retrieve history for
+   * @return {Promise} an object that contains intervals of the user's period (start & length) in that year in reverse chronological order
+   */
+  GETCycleHistoryByYear: async function(year) {
+    let intervals = []
+    let endOfYear = new Date(year,11,31);
+    let isYearsLastPeriod = true;
+
+    let calendar = await getCalendarByYear(year);
+    let periods = await getPeriodsInYear(year, calendar);
+
+    if(periods.length == 0){
+      return intervals;
+    }
+
+
+
+    let current = endOfYear;
+    try {
+      // Search backwards until date switches to the previous year
+      var periodEnd = null;
+      var periodStart = null;
+
+      var isPeriodEnd = false;
+      var isLastPeriodStart = true;
+
+      for (let i =periods.length - 1; i>= 0; i--){
+
+
+        let current = periods[i];
+        if(await isPeriodStart(current, calendar)){
+          if (isPeriodEnd){
+            //handle the case where a period is a single day long
+            periodEnd = current;
+          }
+          periodStart = current;
+          isPeriodEnd = true;
+
+          if(isLastPeriodStart){
+            //handle special case for the last period, since it could possibly span multiple years
+            periodStart = periods[i];
+            periodEnd = await getLastPeriodsEnd(periodStart);
+            var periodDays = getDaysDiffInclusive(periodEnd, periodStart);
+            intervals.push({"start": periodStart, "periodDays": periodDays});
+          }
+          else if(periodEnd && periodStart){
+            //general case for any period besides first or last in the year
+            var periodDays = getDaysDiffInclusive(periodEnd, periodStart);
+            intervals.push({"start": periodStart, "periodDays": periodDays});
+          }
+          isLastPeriodStart = false;
+        }
+        else{
+          if(isPeriodEnd){
```

Comment:
nice! if X is first X seen after an interval, label it as the end of the period 

good for cases such as _ _ X X X _ X _ X _ _